src/server/wesnothd/server.cpp

```diff
@@ -1742,7 +1742,22 @@ void server::handle_player_in_game(player_iterator p, simple_wml::document& data
 						source = "Default";
 					}
 				}
-				user_handler_->db_insert_game_player_info(uuid_, g.db_id(), side["player_id"].to_string(), side["side"].to_int(), side["is_host"].to_bool(), side["faction"].to_string(), version, source, side["current_player"].to_string());
+
+				// add leader data
+				std::vector<std::string> leaders;
+				if(side.attr("type") != "") {
+					leaders.emplace_back(side.attr("type").to_string());
+				}
+				for(const auto unit : side.children("unit")) {
+					if(unit->attr("canrecruit") == "yes") {
+						leaders.emplace_back(unit->attr("type").to_string());
+					}
+				}
+				for(const auto leader : side.children("leader")) {
+					leaders.emplace_back(leader->attr("type").to_string());
+				}
+
+				user_handler_->db_insert_game_player_info(uuid_, g.db_id(), side["player_id"].to_string(), side["side"].to_int(), side["is_host"].to_bool(), side["faction"].to_string(), version, source, side["current_player"].to_string(), utils::join(leaders));
```

Comment:
This doesn't look reversible if a unit ID contains a comma. Is the whole list even that useful? Maybe just store the first one?

Reply:
I mean utils::join(), as far as I can tell it expects that the separator (comma by default) to just not be in the list. If a leader unit ID contains a comma then trying to split on comma to get the list back will not work.

Reply:
The wiki says:
> An id should consist only of alphanumerics and spaces (or underscores).
> ...
> WARNING : characters "$", "&", "*", "-", "(" and ")" are illegal for use in the unit type id and must be avoided because they might lead to corrupted saved games.

Does that mean that only alphanumeric, spaces, and underscores are allowed, or that anything is technically allowed aside from the ones specifically in the warning? I was thinking it was the former, so commas in a unit type wouldn't occur.

Reply:
I doubt that warning is even uptodate. It's probably from when unit type IDs where used as WML keys in saves/replays.

Turns out there is enforcing of the suggestion from the wiki since a while. (The function name is misleading it actually replaces invalid characters with underscores.)
https://github.com/wesnoth/wesnoth/blob/e0e8af894eb877c1017165b512e1a039528978e4/src/units/types.cpp#L1437
So assuming that function is used consistently I withdraw my comment. The wiki should be updated.